### ISSUE TITLE
Packages with meta.license as string are always considered as free

### ISSUE THREAD

**Issue by adisbladis** [NixOS/nixpkgs#269788]
### Describe the bug
Due to a logic bug in `check-meta.nix` all packages where `meta.license` is a bare string are always considered to be free.
We should consider all licenses that we do not know to be free to be unfree.

The current implementation looks like:
``` nix
isUnfree = licenses: lib.lists.any (l: !l.free or true) licenses;
```
which always evaluates to `!true` for strings.

### Steps To Reproduce
1. `nix-instantiate ./. -A boehmgc`

### Expected behavior
All unknown licenses are considered unfree.

### Proposed solution
I suggest that we fix the issue by first making all `meta.license` fields structured on all packages.
Once that's done we can change `isUnfree` from `check-meta.nix` to correctly return `true` for bare strings. 

### Methodology
I've dumped the script I've used to find packages with bare strings as meta.license in https://gist.github.com/adisbladis/b5aa64ded97712e062728bd033e90fa2.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

### Priorities

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by aidalgol** [comment]
Created #269789 to address this in `teensy-udev-rules`.

**Comment by lourkeur** [comment]
teck-programmer and teck-udev-rules should be done with #269855

**Comment by ncfavier** [comment]
Maybe you could explain what should be done with stringy licenses, like
- `teeworlds`'s ``"BSD-style, see `license.txt'"``
- `lambdabot`'s (really, `cabal2nix`'s) `"GPL"`

**Comment by AndersonTorres** [comment]
> Maybe you could explain what should be done with stringy licenses

Convert them to not-so-stringly ones.

**Comment by Artturin** [comment]
License can be a attr set like

```
license = {
  url = "https://raw.githubusercontent.com/m35/jpsxdec/readme/.github/LICENSE.md";
  free = true;
};
```

**Comment by pbsds** [comment]
string licenses should be eval error, with a helpful error message

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/blob/911ce27f7ec3742e396928da3ccb25c17566bdfe/pkgs/stdenv/generic/check-meta.nix#L267-L269

**Comment by AndersonTorres** [comment]
> string licenses should be eval error, with a helpful error message

For now a warning is better.

**Comment by Luflosi** [comment]
Thank you.
`teensy-udev-rules`, `ocamlPackages.caqti`, `ocamlPackages.caqti-async`, `ocamlPackages.caqti-driver-mariadb`, `ocamlPackages.caqti-driver-postgresql`, `ocamlPackages.caqti-dynload`, `ocamlPackages.caqti-lwt`, `ocamlPackages.caqti-driver-sqlite3` and `ocamlPackages.caqti-type-calendar` can also be marked as done.

**Comment by bjornfor** [comment]
> Thank you. `teensy-udev-rules`, `ocamlPackages.caqti`, `ocamlPackages.caqti-async`, `ocamlPackages.caqti-driver-mariadb`, `ocamlPackages.caqti-driver-postgresql`, `ocamlPackages.caqti-dynload`, `ocamlPackages.caqti-lwt`, `ocamlPackages.caqti-driver-sqlite3` and `ocamlPackages.caqti-type-calendar` can also be marked as done.

Marked as done. (Thanks!)

**Comment by Luflosi** [comment]
`postcss-cli` can also be checked off.

**Comment by SuperSandro2000** [comment]
> For now a warning is better.

which gets treated by ofborg as a hard failure....

**Comment by AndersonTorres** [comment]
> which gets treated by ofborg as a hard failure....

Even in old code?
Not that I think it is a problem at all. The monorepo can be stricter than a local instance.

**Comment by SuperSandro2000** [comment]
> Even in old code?

Yes. You cannot yield any lib.warn in the paths that ofborg checks. If the code is technically unused within nixpkgs from the things ofborg checks and may only be triggered by out of tree things, then it is fine.

**Comment by necessarily-equal** [comment]
`bookletimposer` can be handled by deleting the package https://github.com/NixOS/nixpkgs/pull/278100

**Comment by jopejoe1** [comment]
Caught a few more stragglers: #512729, #512727, and #512689. After those, only https://github.com/NixOS/cabal2nix/pull/677 is left to fix the Haskell package set; I would hope for some reviews on those.
